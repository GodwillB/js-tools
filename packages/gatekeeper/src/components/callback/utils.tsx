import { User } from '@onaio/session-reducer';
import {
  AN_ERROR_OCCURRED,
  NOTHING_HERE,
  PLEASE_WAIT,
  WELCOME_BACK
} from 'gatekeeper/src/helpers/constants';
import React from 'react';

/** describing props for the util components */
interface Props {
  message?: string;
}

/** default 404 page component */
export const Component404 = ({ message = NOTHING_HERE }: Props) => {
  return (
    <div className="gatekeeper-cb">
      <p className="gatekeeper-p">{message}</p>
    </div>
  );
};

/** error page component */
export const RenderErrorComponent = ({ message = AN_ERROR_OCCURRED }: Props) => {
  return (
    <div className="gatekeeper-cb">
      <p className="gatekeeper-p">{message}</p>
    </div>
  );
};

/** loading component */
export const RenderLoadingComponent = ({ message = PLEASE_WAIT }: Props) => {
  return (
    <div className="gatekeeper-cb">
      <p className="gatekeeper-p">{message}</p>
    </div>
  );
};

/** interface for SuccessfulLogin props */
export interface SuccessfulLoginProps {
  extraData?: { [key: string]: any } /** can be an object with any properties */;
  user: User;
}

/** successful login page component */
export const SuccessfulLogin = (props: SuccessfulLoginProps) => {
  const { user } = props;
  return (
    <div className="gatekeeper-cb">
      <p className="gatekeeper-p">
        `${WELCOME_BACK}, {user.username}!`
      </p>
    </div>
  );
};
