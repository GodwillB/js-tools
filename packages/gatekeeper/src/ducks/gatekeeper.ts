import { ActionCreator, AnyAction, Store } from 'redux';
import SeamlessImmutable from 'seamless-immutable';

export const reducerName = 'gatekeeper';

/** Interface for Record action */
export interface RecordAction extends AnyAction {
  result: { [key: string]: any };
  success: boolean | null;
  type: typeof RECORD;
}

/** interface to describe GateKeeper state */
export interface GateKeeperState {
  result: { [key: string]: any } /** stores the result of the auth attempt */;
  success: boolean | null /** was it successful or not */;
}

/** immutable GateKeeper state */
export type ImmutableGateKeeperState = GateKeeperState &
  SeamlessImmutable.ImmutableObject<GateKeeperState>;

/** Initial state for GateKeeper */
export const initialState: ImmutableGateKeeperState = SeamlessImmutable({
  result: {},
  success: null
});

/** GateKeeper reducer function
 * @param {initialState} state - the initial state
 * @param {AnyAction} action - the action
 */
export default function reducer(state = initialState, action: AnyAction): ImmutableGateKeeperState {
  switch (action.type) {
    case RECORD:
      return state.merge({
        result: { ...action.result },
        success: action.success
      });
    default:
      return state;
  }
}

// actions

/** authenticate success action type */
export const RECORD = '@onaio/gatekeeper/reducer/RECORD';

// action creators
/** record the result of the authentication attempt
 * @param {boolean} success - whether it was successful or not
 * @param {{ [key: string]: any }} result - an object containing result information
 */
export const recordResult: ActionCreator<RecordAction> = (
  success: boolean,
  result: { [key: string]: any } = {}
) => ({
  result,
  success,
  type: RECORD
});

// selectors

/** get result
 * @param {Partial<Store>} state - the redux store
 */
export function getResult(state: Partial<Store>): { [key: string]: any } {
  return (state as any)[reducerName].result;
}

/** get success
 * @param {Partial<Store>} state - the redux store
 */
export function getSuccess(state: Partial<Store>): boolean {
  return (state as any)[reducerName].success;
}
