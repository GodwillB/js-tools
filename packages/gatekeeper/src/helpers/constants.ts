export const ONADATA_PROVIDER_NAME = 'onadata';
export type ONADATA_PROVIDER_NAME = typeof ONADATA_PROVIDER_NAME;
