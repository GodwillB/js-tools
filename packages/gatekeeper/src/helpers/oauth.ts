import ClientOAuth2 from 'client-oauth2';

export const onadataAuth = new ClientOAuth2({
  accessTokenUri: 'https://stage-api.ona.io/o/token/',
  authorizationUri: 'https://stage-api.ona.io/o/authorize/',
  clientId: 'iDWxqlTdUNUtGd3MIjxsI8HcICSjsRwHfy1LKvzF',
  redirectUri: 'http://localhost:3000/oauth/callback/onadata/',
  scopes: ['read', 'write'],
  state: 'abc'
});
