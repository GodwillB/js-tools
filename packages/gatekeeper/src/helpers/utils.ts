/** Type definition for error callback function  */
export type ErrorCallback = (e: string) => string | void;

/** Error callback function
 * @param {string} error - the error message
 */
export function errorCallback(error: string) {
  throw new Error(error);
}
