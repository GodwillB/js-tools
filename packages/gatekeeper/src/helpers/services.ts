import { AuthenticateAction, authenticateUser } from '@onaio/session-reducer';
import ClientOAuth2 from 'client-oauth2';
import { Dispatch } from 'react';
import { ActionCreator } from 'redux';
import { getOnadataUserInfo, UserInfoFnType } from './oauth';
import { ErrorCallback, errorCallback } from './utils';

/** Calls the oAuth provider to get user details
 * @param {string} locationHash - the location hash value that we receive from the oAuth provider
 * @param {string} url - the URL that returns the user information for the provider
 * @param {ClientOAuth2} provider - the Oauth client object for the provider
 * @param {UserInfoFnType} userInfoCallback - function the gets user info from API response
 */
export async function oauth2Callback(
  locationHash: string,
  url: string,
  provider: ClientOAuth2,
  userInfoCallback: UserInfoFnType
) {
  return provider.token.getToken(locationHash).then(async (oAuthObject: any) => {
    const response = await fetch(
      url,
      oAuthObject.sign({
        method: 'GET',
        url
      })
    );

    if (!response.ok) {
      throw new Error(`oAuth service oauth2Callback failed, HTTP status ${response.status}`);
    }

    const data = await response.json();

    return userInfoCallback(data);
  });
}

/** This function is used to fetch the user logging in by calling oauth2Callback
 * and then calling authenticateUser to store the user in the Redux store.
 * @param {string} locationHash - the location hash value that we receive from the oAuth provider
 * @param {string} url - the URL that returns the user information for the provider
 * @param {any} dispatch - the dispatch function that calls the session reducer action creator
 * @param {ClientOAuth2} provider - the Oauth client object for the provider
 * @param {UserInfoFnType} userInfoCallback - function the gets user info from API response
 * @param {ErrorCallback} errorCallbackFn - a function that handles error messages
 */
export async function fetchUser(
  locationHash: string,
  url: string,
  dispatch: Dispatch<any>,
  provider: ClientOAuth2,
  authenticateActionCreator: ActionCreator<AuthenticateAction> = authenticateUser,
  userInfoCallback: UserInfoFnType = getOnadataUserInfo,
  errorCallbackFn: ErrorCallback = errorCallback
) {
  try {
    const userInfo = await oauth2Callback(locationHash, url, provider, userInfoCallback);
    if (userInfo) {
      const { authenticated, user, extraData } = userInfo;
      dispatch(authenticateActionCreator(authenticated, user, extraData));
    } else {
      errorCallbackFn('Something went wrong');
    }
  } catch (error) {
    errorCallbackFn(error.message);
  }
}
