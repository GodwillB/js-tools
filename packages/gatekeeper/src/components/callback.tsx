import {
  authenticateUser,
  getExtraData,
  getUser,
  initialState,
  isAuthenticated,
  User
} from '@onaio/session-reducer';
import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { RouteComponentProps } from 'react-router';
import { getProviderFromOptions, Providers } from '../helpers/oauth';
import { fetchUser } from '../helpers/services';

/** Route params interface */
export interface RouteParams {
  id: string;
}

/** interface for OauthCallbackProps props */
export interface OauthCallbackProps<G> extends RouteComponentProps<G> {
  ErrorPageComponent?: React.ElementType;
  HTTP404Component?: React.ElementType;
  SuccessfulLoginComponent?: React.ElementType;
  providers: Providers;
}

/** default 404 page component */
const Component404 = () => {
  return (
    <div className="gatekeeper-cb">
      <p className="gatekeeper-p">Nothing here!</p>
    </div>
  );
};

/** error page component */
const ErrorPage = () => {
  return (
    <div className="gatekeeper-cb">
      <p className="gatekeeper-p">An error occurred!</p>
    </div>
  );
};

/** interface for SuccessfulLogin props */
interface SuccessfulLoginProps {
  user: User;
}

/** successful login page component */
const SuccessfulLogin = (props: SuccessfulLoginProps) => {
  const { user } = props;
  return (
    <div className="gatekeeper-cb">
      <p className="gatekeeper-p">Welcome back, {user.username}!</p>
    </div>
  );
};

/** The oAuth callback component
 * This component should be on the page that receives the callback from the
 * oAuth provider.
 * It attempts to get the oAuth provider from the URL and then process the oAuth
 * login attempt.  This uses react-router
 * For instance, if you have a provider named onadata, this component should be
 * on a page that matches this pattern "https://example.com/callback/onadata"
 * Once successfully processed, the user is stored in the session Reducer.
 */
const OauthCallback = (props: any) => {
  const {
    HTTP404Component,
    SuccessfulLoginComponent,
    authenticated,
    providers,
    sessionData,
    sessionUser
  } = props;
  const locationHash = props.location.hash;
  const id = props.match.params.id;

  if (!Object.keys(providers).includes(id) && HTTP404Component) {
    return <HTTP404Component />;
  }

  const providerOptions = providers[id];
  const { userUri } = providerOptions;
  const provider = getProviderFromOptions(providerOptions);

  useEffect(() => {
    if (authenticated === false) {
      fetchUser(locationHash, userUri, provider, props.authenticateUser);
    }
  }, []); // The empty array causes this effect to only run on mount

  const successProps = { user: sessionUser };

  if (authenticated && SuccessfulLoginComponent) {
    return SuccessfulLoginComponent && <SuccessfulLoginComponent {...successProps} />;
  }

  return <div>no dice!</div>;
};

OauthCallback.defaultProps = {
  ErrorPageComponent: ErrorPage,
  HTTP404Component: Component404,
  SuccessfulLoginComponent: SuccessfulLogin
};

export { OauthCallback }; // export the un-connected component

/** Connect the component to the store */

const mapStateToProps = (state: any) => {
  return {
    authenticated: isAuthenticated(state),
    sessionData: getExtraData(state),
    sessionUser: getUser(state)
  };
};

const mapDispatchToProps = { authenticateUser };

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(OauthCallback);
