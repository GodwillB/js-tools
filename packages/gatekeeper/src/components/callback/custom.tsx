import {
  AuthenticateAction,
  authenticateUser,
  getExtraData,
  getUser,
  isAuthenticated
} from '@onaio/session-reducer';
import React, { useEffect } from 'react';
import { ActionCreator, Store } from 'redux';
import { getSuccess, RecordAction, recordResult } from '../../ducks/gatekeeper';
import { fetchState } from '../../helpers/services';
import { errorCallback } from '../../helpers/utils';
import {
  BaseCallbackComponent,
  BaseCallbackComponentProps,
  defaultBaseCallbackComponentProps
} from './base';
import { Component404, RenderErrorComponent } from './utils';

interface RouteParams {
  id: string;
}

/** interface for OauthCallbackProps props */
export interface OauthCallbackProps<G> extends BaseCallbackComponentProps<G> {
  ErrorComponent: React.ElementType;
  HTTP404Component: React.ElementType;
  apiURL: string;
  authenticateActionCreator: ActionCreator<AuthenticateAction>;
  recordResultActionCreator: ActionCreator<RecordAction>;
}

/** default props for OauthCallback */
export const defaultOauthCallbackProps: Partial<OauthCallbackProps<RouteParams>> = {
  ...defaultBaseCallbackComponentProps,
  ErrorComponent: RenderErrorComponent,
  HTTP404Component: Component404,
  UnSuccessfulLoginComponent: RenderErrorComponent,
  authenticateActionCreator: authenticateUser,
  recordResultActionCreator: recordResult
};

const CustomCallback = (props: OauthCallbackProps<RouteParams>) => {
  const {
    LoadingComponent,
    SuccessfulLoginComponent,
    UnSuccessfulLoginComponent,
    apiURL,
    authSuccess,
    authenticated,
    authenticateActionCreator,
    recordResultActionCreator,
    sessionData,
    sessionUser
  } = props;

  useEffect(() => {
    if (authSuccess === null || authenticated === false) {
      fetchState(apiURL, authenticateActionCreator, recordResultActionCreator, errorCallback).catch(
        e => {
          /** do nothing - is this wise?? */
        }
      );
    }
  }, []); // The empty array causes this effect to only run on mount

  const baseProps = {
    LoadingComponent,
    SuccessfulLoginComponent,
    UnSuccessfulLoginComponent,
    authSuccess,
    authenticated,
    sessionData,
    sessionUser
  };

  return <BaseCallbackComponent {...baseProps} />;
};

CustomCallback.defaultProps = defaultOauthCallbackProps;

export { CustomCallback };
