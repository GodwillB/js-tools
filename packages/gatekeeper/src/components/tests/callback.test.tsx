import store from '@onaio/connected-reducer-registry';
import reducerRegistry from '@onaio/redux-reducer-registry';
import session, {
  authenticateUser,
  logOutUser,
  reducerName as sessionReducer
} from '@onaio/session-reducer';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import fetchMock from 'fetch-mock';
import { createBrowserHistory } from 'history';
import React from 'react';
import { Provider } from 'react-redux';
import gatekeeper, { recordResult, reducerName as gateKeeperReducer } from '../../ducks/gatekeeper';
import { getOnadataUserInfo } from '../../helpers/oauth';
import * as serviceHelpers from '../../helpers/services';
import * as helperFixtures from '../../helpers/tests/fixtures';
import * as callback from '../callback';
import * as fixtures from './fixtures';

const ConnectedOauthCallback = callback.default;
const OauthCallback = callback.OauthCallback;

const history = createBrowserHistory();
reducerRegistry.register(sessionReducer, session);
reducerRegistry.register(gateKeeperReducer, gatekeeper);

describe('gatekeeper/OauthCallback', () => {
  beforeEach(() => {
    jest.resetAllMocks();
    fetchMock.restore();
  });

  it('renders without crashing', () => {
    fetchMock.getOnce(
      fixtures.providers.onadata.userUri,
      JSON.stringify(helperFixtures.onadataUser)
    );
    const props = {
      authSuccess: true,
      authenticateActionCreator: authenticateUser,
      authenticated: false,
      history,
      location: {
        hash:
          '#access_token=iLoveOov&expires_in=36000&token_type=Bearer&scope=read+write&state=abc',
        pathname: '/oauth/callback/404/',
        search: '',
        state: undefined
      },
      match: {
        isExact: true,
        params: { id: '404' },
        path: 'https://example.com/oauth/callback/:id',
        url: 'https://example.com/oauth/callback/404/'
      },
      providers: fixtures.providers,
      sessionData: {},
      sessionUser: {
        email: '',
        gravatar: '',
        name: '',
        username: ''
      }
    };
    shallow(<OauthCallback {...props} />);
  });

  it('renders correctly when not loading', () => {
    fetchMock.getOnce(
      fixtures.providers.onadata.userUri,
      JSON.stringify(helperFixtures.onadataUser)
    );
    const props = {
      authSuccess: null,
      authenticateActionCreator: authenticateUser,
      authenticated: false,
      history,
      location: {
        hash:
          '#access_token=iLoveOov&expires_in=36000&token_type=Bearer&scope=read+write&state=abc',
        pathname: '/oauth/callback/onadata/',
        search: '',
        state: undefined
      },
      match: {
        isExact: true,
        params: { id: 'onadata' },
        path: 'https://example.com/oauth/callback/:id',
        url: 'https://example.com/oauth/callback/onadata/'
      },
      providers: fixtures.providers,
      sessionData: {},
      sessionUser: {
        email: '',
        gravatar: '',
        name: '',
        username: ''
      }
    };

    store.dispatch(logOutUser());
    const wrapper = mount(<OauthCallback {...props} />);
    expect(toJson(wrapper.find('OauthCallback'))).toMatchSnapshot();
    wrapper.unmount();
  });

  it('renders correctly when not logged in', () => {
    fetchMock.getOnce(
      fixtures.providers.onadata.userUri,
      JSON.stringify(helperFixtures.onadataUser)
    );
    const props = {
      authSuccess: true,
      authenticateActionCreator: authenticateUser,
      authenticated: false,
      history,
      location: {
        hash:
          '#access_token=iLoveOov&expires_in=36000&token_type=Bearer&scope=read+write&state=abc',
        pathname: '/oauth/callback/onadata/',
        search: '',
        state: undefined
      },
      match: {
        isExact: true,
        params: { id: 'onadata' },
        path: 'https://example.com/oauth/callback/:id',
        url: 'https://example.com/oauth/callback/onadata/'
      },
      providers: fixtures.providers,
      sessionData: {},
      sessionUser: {
        email: '',
        gravatar: '',
        name: '',
        username: ''
      }
    };

    store.dispatch(logOutUser());
    const mock = jest.spyOn(serviceHelpers, 'fetchUser');

    const wrapper = mount(<OauthCallback {...props} />);

    const { onadataAuth } = helperFixtures;
    const url = 'https://stage-api.ona.io/api/v1/user.json';
    const hash =
      '#access_token=iLoveOov&expires_in=36000&token_type=Bearer&scope=read+write&state=abc';

    expect(mock).toHaveBeenCalledWith(
      hash,
      url,
      onadataAuth,
      authenticateUser,
      recordResult,
      getOnadataUserInfo
    );

    expect(toJson(wrapper.find('OauthCallback'))).toMatchSnapshot();
    mock.mockRestore();
    wrapper.unmount();
  });

  it('renders correctly when logged in', () => {
    fetchMock.getOnce(
      fixtures.providers.onadata.userUri,
      JSON.stringify(helperFixtures.onadataUser)
    );
    const props = {
      history,
      location: {
        hash:
          '#access_token=iLoveOov&expires_in=36000&token_type=Bearer&scope=read+write&state=abc',
        pathname: '/oauth/callback/onadata/',
        search: '',
        state: undefined
      },
      match: {
        isExact: true,
        params: { id: 'onadata' },
        path: 'https://example.com/oauth/callback/:id',
        url: 'https://example.com/oauth/callback/onadata/'
      },
      providers: fixtures.providers
    };

    const { authenticated, user, extraData } = helperFixtures.onadataSessionWithOauthData;
    store.dispatch(authenticateUser(authenticated, user, extraData));
    store.dispatch(recordResult(true, extraData));

    const wrapper = mount(
      <Provider store={store}>
        <ConnectedOauthCallback {...props} />
      </Provider>
    );

    expect(toJson(wrapper.find('OauthCallback'))).toMatchSnapshot();
    wrapper.unmount();
  });

  it('renders correctly when oAuth error occurred', () => {
    const props = {
      authSuccess: false,
      authenticateActionCreator: authenticateUser,
      authenticated: false,
      history,
      location: {
        hash: '',
        pathname: '/oauth/callback/onadata/',
        search: '?error=unauthorized_client&state=abc',
        state: undefined
      },
      match: {
        isExact: true,
        params: { id: 'onadata' },
        path: 'https://example.com/oauth/callback/:id',
        url: 'https://example.com/oauth/callback/onadata/'
      },
      providers: fixtures.providers,
      sessionData: {},
      sessionUser: {
        email: '',
        gravatar: '',
        name: '',
        username: ''
      }
    };

    const wrapper = mount(<OauthCallback {...props} />);
    expect(toJson(wrapper.find('OauthCallback'))).toMatchSnapshot();
    wrapper.unmount();
  });

  it('renders correctly when uri cant be processed', () => {
    const props = {
      authSuccess: false,
      authenticateActionCreator: authenticateUser,
      authenticated: false,
      history,
      location: {
        hash: '',
        pathname: '/oauth/callback/onadata/',
        search: '',
        state: undefined
      },
      match: {
        isExact: true,
        params: { id: 'onadata' },
        path: 'https://example.com/oauth/callback/:id',
        url: 'https://example.com/oauth/callback/onadata/'
      },
      providers: fixtures.providers,
      sessionData: {},
      sessionUser: {
        email: '',
        gravatar: '',
        name: '',
        username: ''
      }
    };

    const wrapper = mount(<OauthCallback {...props} />);
    expect(toJson(wrapper.find('OauthCallback'))).toMatchSnapshot();
    wrapper.unmount();
  });

  it('renders correctly when provider is not known', () => {
    const props = {
      authenticateActionCreator: authenticateUser,
      authenticated: false,
      history,
      location: {
        hash: '',
        pathname: '/oauth/callback/unknown/',
        search: '',
        state: undefined
      },
      match: {
        isExact: true,
        params: { id: 'unknown' },
        path: 'https://example.com/oauth/callback/:id',
        url: 'https://example.com/oauth/callback/unknown/'
      },
      providers: fixtures.providers,
      sessionData: {},
      sessionUser: {
        email: '',
        gravatar: '',
        name: '',
        username: ''
      }
    };

    const wrapper = mount(<OauthCallback {...props} />);
    expect(toJson(wrapper.find('OauthCallback'))).toMatchSnapshot();
    wrapper.unmount();
  });
});
