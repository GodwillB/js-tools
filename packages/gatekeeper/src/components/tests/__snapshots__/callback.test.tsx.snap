// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gatekeeper/OauthLogin renders correctly when logged in 1`] = `
<OauthCallback
  ErrorComponent={[Function]}
  HTTP404Component={[Function]}
  SuccessfulLoginComponent={[Function]}
  UnSuccessfulLoginComponent={[Function]}
  authenticateActionCreator={[Function]}
  authenticated={true}
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  location={
    Object {
      "hash": "#access_token=iLoveOov&expires_in=36000&token_type=Bearer&scope=read+write&state=abc",
      "pathname": "/oauth/callback/onadata/",
      "search": "",
      "state": undefined,
    }
  }
  match={
    Object {
      "isExact": true,
      "params": Object {
        "id": "onadata",
      },
      "path": "https://example.com/oauth/callback/:id",
      "url": "https://example.com/oauth/callback/onadata/",
    }
  }
  providers={
    Object {
      "onadata": Object {
        "accessTokenUri": "https://stage-api.ona.io/o/token/",
        "authorizationUri": "https://stage-api.ona.io/o/authorize/",
        "clientId": "hunter2",
        "redirectUri": "https://example.com/oauth/callback/onadata/",
        "scopes": Array [
          "read",
          "write",
        ],
        "state": "abc",
        "userUri": "https://stage-api.ona.io/api/v1/user.json",
      },
    }
  }
  sessionData={
    Object {
      "api_token": "the api token",
      "city": "",
      "country": "",
      "email": "mosh@example.com",
      "gravatar": "https://secure.gravatar.com/avatar/ae22ab897231db07205bd5d00e64cbbf?d=https%3A%2F%2Fona.io%2Fstatic%2Fimages%2Fdefault_avatar.png&s=60",
      "name": "mosh",
      "organization": "",
      "require_auth": true,
      "temp_token": "the temp token",
      "twitter": "",
      "url": "https://stage-api.ona.io/api/v1/profiles/moshthepitt",
      "user": "https://stage-api.ona.io/api/v1/users/moshthepitt",
      "username": "moshthepitt",
      "website": "",
    }
  }
  sessionUser={
    Object {
      "email": "mosh@example.com",
      "gravatar": "https://secure.gravatar.com/avatar/ae22ab897231db07205bd5d00e64cbbf?d=https%3A%2F%2Fona.io%2Fstatic%2Fimages%2Fdefault_avatar.png&s=60",
      "name": "mosh",
      "username": "moshthepitt",
    }
  }
>
  <SuccessfulLogin
    extraData={
      Object {
        "api_token": "the api token",
        "city": "",
        "country": "",
        "email": "mosh@example.com",
        "gravatar": "https://secure.gravatar.com/avatar/ae22ab897231db07205bd5d00e64cbbf?d=https%3A%2F%2Fona.io%2Fstatic%2Fimages%2Fdefault_avatar.png&s=60",
        "name": "mosh",
        "organization": "",
        "require_auth": true,
        "temp_token": "the temp token",
        "twitter": "",
        "url": "https://stage-api.ona.io/api/v1/profiles/moshthepitt",
        "user": "https://stage-api.ona.io/api/v1/users/moshthepitt",
        "username": "moshthepitt",
        "website": "",
      }
    }
    user={
      Object {
        "email": "mosh@example.com",
        "gravatar": "https://secure.gravatar.com/avatar/ae22ab897231db07205bd5d00e64cbbf?d=https%3A%2F%2Fona.io%2Fstatic%2Fimages%2Fdefault_avatar.png&s=60",
        "name": "mosh",
        "username": "moshthepitt",
      }
    }
  >
    <div
      className="gatekeeper-cb"
    >
      <p
        className="gatekeeper-p"
      >
        Welcome back, 
        moshthepitt
        !
      </p>
    </div>
  </SuccessfulLogin>
</OauthCallback>
`;

exports[`gatekeeper/OauthLogin renders correctly when not logged in 1`] = `
<OauthCallback
  ErrorComponent={[Function]}
  HTTP404Component={[Function]}
  SuccessfulLoginComponent={[Function]}
  UnSuccessfulLoginComponent={[Function]}
  authenticateActionCreator={[Function]}
  authenticated={false}
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  location={
    Object {
      "hash": "#access_token=iLoveOov&expires_in=36000&token_type=Bearer&scope=read+write&state=abc",
      "pathname": "/oauth/callback/onadata/",
      "search": "",
      "state": undefined,
    }
  }
  match={
    Object {
      "isExact": true,
      "params": Object {
        "id": "onadata",
      },
      "path": "https://example.com/oauth/callback/:id",
      "url": "https://example.com/oauth/callback/onadata/",
    }
  }
  providers={
    Object {
      "onadata": Object {
        "accessTokenUri": "https://stage-api.ona.io/o/token/",
        "authorizationUri": "https://stage-api.ona.io/o/authorize/",
        "clientId": "hunter2",
        "redirectUri": "https://example.com/oauth/callback/onadata/",
        "scopes": Array [
          "read",
          "write",
        ],
        "state": "abc",
        "userUri": "https://stage-api.ona.io/api/v1/user.json",
      },
    }
  }
  sessionData={Object {}}
  sessionUser={
    Object {
      "email": "",
      "gravatar": "",
      "name": "",
      "username": "",
    }
  }
>
  <RenderErrorComponent>
    <div
      className="gatekeeper-cb"
    >
      <p
        className="gatekeeper-p"
      >
        An error occurred!
      </p>
    </div>
  </RenderErrorComponent>
</OauthCallback>
`;

exports[`gatekeeper/OauthLogin renders correctly when oAuth error occurred 1`] = `
<OauthCallback
  ErrorComponent={[Function]}
  HTTP404Component={[Function]}
  SuccessfulLoginComponent={[Function]}
  UnSuccessfulLoginComponent={[Function]}
  authenticateActionCreator={[Function]}
  authenticated={false}
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  location={
    Object {
      "hash": "",
      "pathname": "/oauth/callback/onadata/",
      "search": "?error=unauthorized_client&state=abc",
      "state": undefined,
    }
  }
  match={
    Object {
      "isExact": true,
      "params": Object {
        "id": "onadata",
      },
      "path": "https://example.com/oauth/callback/:id",
      "url": "https://example.com/oauth/callback/onadata/",
    }
  }
  providers={
    Object {
      "onadata": Object {
        "accessTokenUri": "https://stage-api.ona.io/o/token/",
        "authorizationUri": "https://stage-api.ona.io/o/authorize/",
        "clientId": "hunter2",
        "redirectUri": "https://example.com/oauth/callback/onadata/",
        "scopes": Array [
          "read",
          "write",
        ],
        "state": "abc",
        "userUri": "https://stage-api.ona.io/api/v1/user.json",
      },
    }
  }
  sessionData={Object {}}
  sessionUser={
    Object {
      "email": "",
      "gravatar": "",
      "name": "",
      "username": "",
    }
  }
>
  <RenderErrorComponent>
    <div
      className="gatekeeper-cb"
    >
      <p
        className="gatekeeper-p"
      >
        An error occurred!
      </p>
    </div>
  </RenderErrorComponent>
</OauthCallback>
`;

exports[`gatekeeper/OauthLogin renders correctly when uri cant be processed 1`] = `
<OauthCallback
  ErrorComponent={[Function]}
  HTTP404Component={[Function]}
  SuccessfulLoginComponent={[Function]}
  UnSuccessfulLoginComponent={[Function]}
  authenticateActionCreator={[Function]}
  authenticated={false}
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  location={
    Object {
      "hash": "",
      "pathname": "/oauth/callback/onadata/",
      "search": "",
      "state": undefined,
    }
  }
  match={
    Object {
      "isExact": true,
      "params": Object {
        "id": "onadata",
      },
      "path": "https://example.com/oauth/callback/:id",
      "url": "https://example.com/oauth/callback/onadata/",
    }
  }
  providers={
    Object {
      "onadata": Object {
        "accessTokenUri": "https://stage-api.ona.io/o/token/",
        "authorizationUri": "https://stage-api.ona.io/o/authorize/",
        "clientId": "hunter2",
        "redirectUri": "https://example.com/oauth/callback/onadata/",
        "scopes": Array [
          "read",
          "write",
        ],
        "state": "abc",
        "userUri": "https://stage-api.ona.io/api/v1/user.json",
      },
    }
  }
  sessionData={Object {}}
  sessionUser={
    Object {
      "email": "",
      "gravatar": "",
      "name": "",
      "username": "",
    }
  }
>
  <RenderErrorComponent>
    <div
      className="gatekeeper-cb"
    >
      <p
        className="gatekeeper-p"
      >
        An error occurred!
      </p>
    </div>
  </RenderErrorComponent>
</OauthCallback>
`;
